### PR TITLE
Upgrade 1020

### PR DESCRIPTION
**Pull request by mikiquantum**
# Description

Last items to complete before code freeze:

- [x] XCM v3 Migration
    - [x] Asset Registry (Migrate current registry location entry of `CurrencyId::Native`)
    - [x] ...
- [x] Add Liquidity Pool Gateway Pallet to runtime [PR](https://github.com/centrifuge/centrifuge-chain/pull/1523)
    - [x] Dummy Liquidity Pool `InboundQueue` that triggers event with info (Runtime struct)
    - [x] Include Liquidity Pools to runtime
        - NOT use LPs as InboundQueue
    - [x] Wrapper `OutboundQueue` that LPs consumes
        - Only allow (`AddCurrency`, `UpdateMember`, `AddPool`, `AddTranche`, `UpdateTrancheTokenPrice`)
        - Uses gateway as real outbound 
- [x] Custom LP XCM Converter to GatewayOrigin #1521 
  *NOTE: This one will be merged to main in order to allow upgrading DEV earlier*
  - Will be added with [this PR](https://github.com/centrifuge/centrifuge-chain/pull/1525/files)
- [x] Adapt Liquidity Pools Weights [PR](https://github.com/centrifuge/centrifuge-chain/pull/1521)
- [x] Bump versions

### CODE REVIEW

File: runtime/centrifuge/src/migrations.rs [hunk 1]
```diff
@@ -234,4 +248,243 @@ mod asset_registry {
 			Ok(())
 		}
 	}
+
+	pub struct AssetRegistryMultilocationToXCMV3<T>(PhantomData<T>);
```
Comment (mikiquantum) [inline runtime/centrifuge/src/migrations.rs:252]: Conceptually this is it. I will clean it up tomorrow

File: runtime/altair/src/lib.rs [hunk 2]
```diff
@@ -124,7 +123,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("altair"),
 	impl_name: create_runtime_str!("altair"),
 	authoring_version: 1,
-	spec_version: 1030,
+	spec_version: 1031,
```
Comment (mikiquantum) [inline runtime/altair/src/lib.rs:126]: Need to bump bc Algol is already in 30

File: runtime/altair/src/migrations.rs [hunk 3]
```diff
@@ -10,22 +10,27 @@
...
+use xcm::{v3::prelude::*, VersionedMultiLocation};
 
 use crate::{RocksDbWeight, Runtime};
 
-/// The migration set for Altair 1030 @ Kusama. It includes all the migrations
+/// The migration set for Altair 1031 @ Kusama. It includes all the migrations
 /// that have to be applied on that chain, which includes migrations that have
 /// already been executed on Algol (1028 & 1029).
 #[cfg(not(feature = "testnet-runtime"))]
-pub type UpgradeAltair1030 = (
```
Comment (mikiquantum) [inline runtime/altair/src/migrations.rs:26]: Needed to bump bc Algol was already in 1030

File: runtime/altair/src/migrations.rs [hunk 4]
```diff
@@ -35,21 +40,30 @@ pub type UpgradeAltair1030 = (
...
-	runtime_common::migrations::nuke::Migration<crate::Loans, RocksDbWeight, 1>,
-	runtime_common::migrations::nuke::Migration<crate::InterestAccrual, RocksDbWeight, 0>,
-	pallet_rewards::migrations::new_instance::FundExistentialDeposit<
-		crate::Runtime,
-		pallet_rewards::Instance2,
-		crate::NativeToken,
-		crate::ExistentialDeposit,
-	>,
+pub type UpgradeAltair1031 = (
```
Comment (mikiquantum) [inline runtime/altair/src/migrations.rs:56]: Needed to bump bc Algol was already in 1031

File: runtime/altair/src/migrations.rs [hunk 5]
```diff
@@ -35,21 +40,30 @@ pub type UpgradeAltair1030 = (
...
+	// TODO: Verify that these ones can be removed from Algol upgrade, since the first 2 fail
```
Comment (mikiquantum) [inline runtime/altair/src/migrations.rs:57]: Seems that these can be removed, can anyone confirm, they fail anyways?

File: runtime/centrifuge/src/migrations.rs [hunk 6]
```diff
@@ -181,57 +193,413 @@ mod asset_registry {
 		}
 	}
 
-	/// Register the LiquidityPools Wrapped Ethereum USDC
-	pub struct RegisterLpEthUSDC;
```
Comment (mikiquantum) [inline runtime/centrifuge/src/migrations.rs:185]: Removing `RegisterLpEthUSDC` migration and adding it within the `AssetRegistryMultilocationToXCMV3` one

File: runtime/centrifuge/src/migrations.rs [hunk 7]
```diff
@@ -181,57 +193,413 @@ mod asset_registry {
...
+		// Adding LP USDC here
+		(
+			LP_ETH_USDC_CURRENCY_ID,
+			lp_eth_usdc_metadata(
+				LiquidityPoolsPalletIndex::get(),
+				ETHEREUM_MAINNET_CHAIN_ID,
+				ETHEREUM_USDC,
+			),
+		),
+	]
```
Comment (mikiquantum) [inline runtime/centrifuge/src/migrations.rs:604]: Add here any current registered asset that is added between now and the upgrade.

File: runtime/centrifuge/src/migrations.rs [hunk 8]
```diff
@@ -181,57 +193,413 @@ mod asset_registry {
...
+		>: From<AssetMetadata<u128, cfg_types::tokens::CustomMetadata>>,
+	{
 		fn on_runtime_upgrade() -> Weight {
-			use orml_traits::asset_registry::Mutate;
+			if VERSION.spec_version != 1020 {
+				return Weight::zero();
+			}
+
+			let mut meta_count = orml_asset_registry::Metadata::<T>::iter_keys().count() as u32;
+			let is_centrifuge = meta_count == 6;
```
Comment (mikiquantum) [inline runtime/centrifuge/src/migrations.rs:214]: Bump number according to current number of assets on chain.

File: runtime/altair/src/migrations.rs [hunk 9]
```diff
@@ -37,23 +42,32 @@ pub type UpgradeAltair1030 = (
...
+	// runtime_common::migrations::nuke::Migration<crate::Loans, RocksDbWeight, 1>,
+	// runtime_common::migrations::nuke::Migration<crate::InterestAccrual, RocksDbWeight, 0>,
+	// runtime_common::migrations::nuke::Migration<crate::PoolSystem, RocksDbWeight, 0>,
```
Comment (cdamian) [inline runtime/altair/src/migrations.rs:62]: This was added during re-base.

File: runtime/altair/src/migrations.rs [hunk 10]
```diff
@@ -37,23 +42,32 @@ pub type UpgradeAltair1030 = (
...
+	// runtime_common::migrations::nuke::Migration<crate::Investments, RocksDbWeight, 0>,
```
Comment (cdamian) [inline runtime/altair/src/migrations.rs:63]: This was added during re-base.

File: runtime/altair/src/migrations.rs (hunk 4 above)
Comment (wischli) [reply runtime/altair/src/migrations.rs:60]: NOTE: Algol is currently on 1030.

**Review by mustermeiszer — APPROVED** [review]
Re-approving!

**Pull request merged**